algorithms/linfa-clustering/src/optics/algorithm.rs

```diff
@@ -0,0 +1,496 @@
+use crate::optics::analysis::*;
+use crate::optics::errors::Result;
+use crate::optics::hyperparameters::OpticsHyperParams;
+use hnsw::{Hnsw, Params, Searcher};
+use linfa::traits::Transformer;
+use linfa::Float;
+use ndarray::{ArrayView, ArrayView1, Ix1, Ix2};
+use ndarray_stats::DeviationExt;
+use noisy_float::prelude::*;
+use rand_pcg::Pcg64;
+#[cfg(feature = "serde")]
+use serde_crate::{Deserialize, Serialize};
+use space::MetricPoint;
+use std::cmp::{max, Ordering};
+use std::collections::BTreeSet;
+
+/// Implementation of euclidean distance for ndarray
+struct Euclidean<'a, F>(ArrayView1<'a, F>);
+
+impl<F: Float> MetricPoint for Euclidean<'_, F> {
+    fn distance(&self, rhs: &Self) -> u64 {
+        let val = self.0.l2_dist(&rhs.0).unwrap();
+        space::f64_metric(val)
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+#[cfg_attr(
+    feature = "serde",
+    derive(Serialize, Deserialize),
+    serde(crate = "serde_crate")
+)]
+/// OPTICS (Ordering Points To Identify Clustering Structure) is a clustering algorithm that
+/// doesn't explicitly cluster the data but instead creates an "augmented ordering" of the dataset
+/// representing it's density-based clustering structure. This ordering contains information which
+/// is equivalent to the density-based clusterings and can then be used for automatic and
+/// interactive cluster analysis.
+///
+/// OPTICS cluster analysis can be used to derive clusters equivalent to the output of other
+/// clustering algorithms such as DBSCAN. However, due to it's more complicated neighborhood
+/// queries it typically has a higher computational cost than other more specific algorithms.
```

Comment:
can you also add a link here to another resource for people who want to read more (e.g. the original paper or a wikipedia article)